### PR TITLE
[Cache] deprecate all PSR-16 adapters, provide Psr16Cache instead

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

As discussed in https://github.com/symfony/symfony/issues/28918, PSR-16 implementations are now mostly useless: either PSR-6 or contracts' CacheInterface is always a better fit.

Let's deprecate them all but keep only a `Psr16Cache` to turn any PSR-6 implementation to a PSR-16 one.

From the changelog:
  * removed `psr/simple-cache` dependency, run `composer require psr/simple-cache` if you need it
  * deprecated all PSR-16 adapters, use `Psr16Cache` or `Symfony\Contracts\Cache\CacheInterface` implementations instead
 * deprecated `SimpleCacheAdapter`, use `Psr16Adapter` instead
 * deprecated the "Psr\SimpleCache\CacheInterface" / "cache.app.simple" service, use "Symfony\Contracts\Cache\CacheInterface" / "cache.app" instead

### CODE REVIEW

File: src/Symfony/Component/Cache/Adapter/ApcuAdapter.php [hunk 1]
```diff
@@ -11,17 +11,109 @@
...
     /**
      * @throws CacheException if APCu is not enabled
      */
     public function __construct(string $namespace = '', int $defaultLifetime = 0, string $version = null)
     {
-        $this->init($namespace, $defaultLifetime, $version);
+        if (!static::isSupported()) {
+            throw new CacheException('APCu is not enabled');
+        }
+        if ('cli' === \PHP_SAPI) {
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/ApcuAdapter.php]: what about phpdbg ?

File: src/Symfony/Component/Cache/Psr16Cache.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        } catch (SimpleCacheException $e) {
+            throw $e;
+        } catch (Psr6CacheException $e) {
+            throw new InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+        }
+        if (null !== $ttl) {
+            $item->expiresAfter($ttl);
+        }
+
+        return $this->pool->save($item);
```
Comment (stof) [inline src/Symfony/Component/Cache/Psr16Cache.php]: shouldn't this also wrap the Psr6CacheException ?

File: src/Symfony/Component/Cache/Adapter/ApcuAdapter.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Adapter/ApcuAdapter.php]: That's just copy-pasting from the trait, let's not change nor wonder here.

File: src/Symfony/Component/Cache/Psr16Cache.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Psr16Cache.php]: save() is not allowed to throw - cannot happen per the spec.

**Review by Tobion — DISMISSED** [review]

File: src/Symfony/Component/Cache/Simple/Psr6Cache.php [hunk 3]
```diff
@@ -11,218 +11,13 @@
...
+use Symfony\Component\Cache\Psr16Cache;
+
+@trigger_error(sprintf('The "%s" class is deprecated since Symfony 4.3, use "%s" instead.', Psr6Cache::class, Psr16Cache::class), E_USER_DEPRECATED);
 
 /**
- * @author Nicolas Grekas <p@tchwork.com>
+ * @deprecated since Symfony 4.3, use Psr16Cache instead.
  */
-class Psr6Cache implements CacheInterface, PruneableInterface, ResettableInterface
+class Psr6Cache extends Psr16Cache
```
[the remark this answers is not in the document]
Comment (Tobion) [reply src/Symfony/Component/Cache/Simple/Psr6Cache.php]: ah it's just getting renamed

File: src/Symfony/Component/Cache/Simple/Psr6Cache.php [hunk 4]
```diff
@@ -11,218 +11,13 @@
...
-    public function setMultiple($values, $ttl = null)
-    {
-        $valuesIsArray = \is_array($values);
-        if (!$valuesIsArray && !$values instanceof \Traversable) {
-            throw new InvalidArgumentException(sprintf('Cache values must be array or Traversable, "%s" given', \is_object($values) ? \get_class($values) : \gettype($values)));
-        }
-        $items = array();
-
-        try {
-            if (null !== $f = $this->createCacheItem) {
```
Comment (Tobion) [inline src/Symfony/Component/Cache/Simple/Psr6Cache.php]: The problem I have experienced is that in dev mode the cache is decorated with a TraceableCache. This means it loses this optimization and introduces https://github.com/symfony/symfony/issues/28918#issuecomment-432399481 again

Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Simple/Psr6Cache.php]: Good catch, that should be fixed as a bug!

Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Simple/Psr6Cache.php]: fixed in https://github.com/symfony/symfony/pull/29243

**Comment by nicolas-grekas** [comment]
PR is now ready.
From the changelog:
From the changelog:
  * removed `psr/simple-cache` dependency, run `composer require psr/simple-cache` if you need it
  * deprecated all PSR-16 adapters, use `Psr16Cache` or `Symfony\Contracts\Cache\CacheInterface` implementations instead
 * deprecated `SimpleCacheAdapter`, use `Psr16Adapter instead

Status: needs review

**Comment by stof** [comment]
>     * removed `psr/simple-cache` dependency, run `composer require psr/simple-cache` if you need it

isn't this broken ? Our exceptions implement both the PSR-6 and PSR-16 interfaces, and so both are hard requirements.

File: src/Symfony/Component/Cache/Tests/Adapter/SimpleCacheAdapterTest.php [hunk 5]
```diff
@@ -26,6 +26,6 @@ class SimpleCacheAdapterTest extends AdapterTestCase
 
     public function createCachePool($defaultLifetime = 0)
     {
-        return new SimpleCacheAdapter(new Psr6Cache(new FilesystemAdapter()), '', $defaultLifetime);
+        return new SimpleCacheAdapter(new FilesystemCache(), '', $defaultLifetime);
```
Comment (stof) [inline src/Symfony/Component/Cache/Tests/Adapter/SimpleCacheAdapterTest.php]: isn't this reverting your recent change ?

Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Tests/Adapter/SimpleCacheAdapterTest.php]: correct - and the updated test case is now in Psr16AdapterTest

**Comment by nicolas-grekas** [comment]
> Our exceptions implement both the PSR-6 and PSR-16 interfaces, and so both are hard requirements.

Good catch. Fixed by conditionally implementing the PSR-16 one only when it's available.

**Comment by nicolas-grekas** [comment]
Now deprecating the "Psr\SimpleCache\CacheInterface" / "cache.app.simple" service, use "Symfony\Contracts\Cache\CacheInterface" / "cache.app" instead.

**Comment by nicolas-grekas** [comment]
Now rebased, ready.

File: UPGRADE-4.3.md [hunk 6]
```diff
@@ -16,6 +23,7 @@ Config
 FrameworkBundle
 ---------------
 
+<<<<<<< HEAD
```
Comment (sstok) [inline UPGRADE-4.3.md]: Unresolved rebase conflict.

Comment (nicolas-grekas) [reply UPGRADE-4.3.md]: resolved thanks

**Comment by fabpot** [comment]
Looks like we still need a rebase here :)

**Comment by nicolas-grekas** [comment]
oups, rebased.

**Comment by enumag** [comment]
Missing documentation change. This part is now outdated: https://symfony.com/doc/master/components/cache.html#available-simple-cache-psr-16-classes

**Pull request merged**